utility_analysis_new/combiners.py

```diff
@@ -158,49 +142,65 @@ def compute_pmf(self) -> np.ndarray:
         return pmf
 
 
-class PartitionSelectionCombiner(pipeline_dp.Combiner):
+# PartitionSelectionAccumulator = (probabilities, moments). They are mutually
+# exclusive:
+# If len(probabilities) <= MAX_PROBABILITIES_IN_ACCUMULATOR then 'probabilities'
+# are used otherwise 'moments'. For more details see docstring to
+# PartitionSelectionCalculator.
+PartitionSelectionAccumulator = Tuple[Optional[Tuple[float]],
+                                      Optional[SumOfRandomVariablesMoments]]
+
+
+def _merge_partition_selection_accumulators(
+        acc1: PartitionSelectionAccumulator,
+        acc2: PartitionSelectionAccumulator) -> PartitionSelectionAccumulator:
+    probs1, moments1 = acc1
+    probs2, moments2 = acc2
+    if probs1 and probs2 and len(probs1) + len(
+            probs2) <= MAX_PROBABILITIES_IN_ACCUMULATOR:
+        return (probs1 + probs2, None)
+
+    if moments1 is None:
+        moments1 = _probabilities_to_moments(probs1)
+    if moments2 is None:
+        moments2 = _probabilities_to_moments(probs2)
+
+    return (None, moments1 + moments2)
+
+
+class PartitionSelectionCombiner(UtilityAnalysisCombiner):
     """A combiner for utility analysis counts."""
-    AccumulatorType = PartitionSelectionAccumulator
 
     def __init__(self, params: pipeline_dp.combiners.CombinerParams):
         self._params = params
 
-    def create_accumulator(self, data: Tuple[Sized, int]) -> AccumulatorType:
-        """Creates an accumulator for data.
-
-        Args:
-            data is a Tuple containing; 1) a list of the user's contributions
-            for a single partition, and 2) the total number of partitions a user
-            contributed to.
-
-        Returns:
-            An accumulator for computing probability of selecting partition.
-        """
+    def create_accumulator(
+            self, data: Tuple[Sized, int]) -> PartitionSelectionAccumulator:
```

Comment:
Python question: Does it make a performance difference to use `PartitionSelectionAccumulator` instead of Tuple[Optional[Tuple[float]], Optional[SumOfRandomVariablesMoments]] directly here?

Reply:
Type annotations doesn't have any performance impact, as for me `Tuple[Optional[Tuple[float]], Optional[SumOfRandomVariablesMoments]]` is just too complicated :)